core/arch/arm/kernel/secure_partition.c

```diff
@@ -1,50 +1,255 @@
+{
+	struct sp_session *last = NULL;
+	uint32_t id = MIN_ENDPOINT_ID + 1;
+
+	last = TAILQ_LAST(open_sessions, sp_sessions_head);
+	if (last) {
+		/* This value is less likely to be already used */
+		id = last->endpoint_id + 1;
+	}
+
+	return id;
+}
+
+static TEE_Result sp_create_ctx(const TEE_UUID *uuid, struct sp_session *s)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_ctx *utc = NULL;
+
+	/* Register context */
+	utc = calloc(1, sizeof(struct sp_ctx));
+	if (!utc)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	utc->uctx.ts_ctx = &utc->ts_ctx;
+	utc->open_session = s;
+
+	s->ts_sess.ctx = &utc->ts_ctx;
+
+	utc->ts_ctx.uuid = *uuid;
+	res = vm_info_init(&utc->uctx);
+	if (res)
+		goto err;
+
+	set_sp_ctx_ops(&utc->ts_ctx);
+
+	if (res)
+		goto err;
+
+	vm_set_ctx(NULL);
+	return TEE_SUCCESS;
+
+err:
+	s->ts_sess.ctx = NULL;
+	vm_set_ctx(NULL);
+	pgt_flush_ctx(&utc->ts_ctx);
+
+	tee_pager_rem_um_areas(&utc->uctx);
+	vm_info_final(&utc->uctx);
+	free(utc);
 
-static TEE_Result secure_partition_init(void)
+	return res;
+}
+
+static TEE_Result sp_create_session(struct sp_sessions_head *open_sessions,
+				    const TEE_UUID *uuid,
+				    struct sp_session **sess)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_session *s = calloc(1, sizeof(struct sp_session));
+
+	if (!s)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	s->endpoint_id = new_session_id(open_sessions);
+	if (!s->endpoint_id) {
+		res = TEE_ERROR_OVERFLOW;
+		goto out;
+	}
+
+	DMSG("Loading Secure Partition %pUl", (void *)uuid);
+	res = sp_create_ctx(uuid, s);
+
+out:
+	if (res == TEE_SUCCESS) {
+		TAILQ_INSERT_TAIL(open_sessions, s, link);
+		*sess = s;
+	} else {
+		free(s);
+	}
+	return res;
+}
+
+static TEE_Result sp_init_set_registers(struct sp_ctx *ctx)
 {
+	struct thread_ctx_regs *sp_regs = &ctx->sp_regs;
+	uint64_t x0 = sp_regs->x[0];
+
+	/* Clear all addresses except x0, x0 contains the info address. */
+	memset(sp_regs, 0, sizeof(*sp_regs));
+	sp_regs->sp = ctx->uctx.stack_ptr;
+	sp_regs->x[0] = x0;
+	sp_regs->pc = ctx->uctx.entry_func;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sp_open_session(struct sp_session **sess,
+				  struct sp_sessions_head *open_sessions,
+				  const TEE_UUID *uuid)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_session *s = NULL;
+	struct sp_ctx *ctx = NULL;
+
+	if (!find_secure_partition(uuid))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	res = sp_create_session(open_sessions, uuid, &s);
+	if (res != TEE_SUCCESS) {
+		DMSG("init sp session failed 0x%x", res);
+		return res;
+	}
+
+	if (s->ts_sess.ctx)
+		ctx = to_sp_ctx(s->ts_sess.ctx);
+
+	if (!ctx)
+		return TEE_ERROR_TARGET_DEAD;
+	*sess = s;
+
+	ts_push_current_session(&s->ts_sess);
+	/* Load the SP using ldelf. */
+	ldelf_load_ldelf(&ctx->uctx);
+	res = ldelf_init_with_ldelf(&s->ts_sess, &ctx->uctx);
+
+	/*Make the SP ready for its first run*/
+	s->state = sp_idle;
+	s->caller_id = 0;
+	sp_init_info(ctx);
+	sp_init_set_registers(ctx);
```

Comment:
maybe these 4 lines should be skipped if `res != TEE_SUCCESS`

Reply:
done